sydent/http/servlets/lookupv2servlet.py

```diff
@@ -0,0 +1,124 @@
+        self.lookup_pepper = lookup_pepper
+
+    @jsonwrap
+    def render_POST(self, request):
+        """
+        Perform lookups with potentially hashed 3PID details.
+
+        Depending on our response to /hash_details, the client will choose a
+        hash algorithm and pepper, hash the 3PIDs it wants to lookup, and
+        send them to us, along with the algorithm and pepper it used.
+
+        We first check this algorithm/pepper combo matches what we expect,
+        then compare the 3PID details to what we have in the database.
+
+        Params: A JSON object containing the following keys:
+                * 'addresses': List of hashed/plaintext (depending on the
+                               algorithm) 3PID addresses and mediums.
+                * 'algorithm': The algorithm the client has used to process
+                               the 3PIDs.
+                * 'pepper': The pepper the client has attached to the 3PIDs.
+
+        Returns: Object with key 'mappings', which is a dictionary of results
+                 where each result is a key/value pair of what the client sent, and
+                 the matching Matrix User ID that claims to own that 3PID.
+
+                 User IDs for which no mapping is found are omitted.
+        """
+        send_cors(request)
+
+        err, args = get_args(request, ('addresses', 'algorithm', 'pepper'))
+        if err:
+            return err
+
+        addresses = args['addresses']
+        if not isinstance(addresses, list):
+            request.setResponseCode(400)
+            return {'errcode': 'M_INVALID_PARAM', 'error': 'addresses must be a list'}
+
+        algorithm = str(args['algorithm'])
+        if algorithm not in HashDetailsServlet.known_algorithms:
+            request.setResponseCode(400)
+            return {'errcode': 'M_INVALID_PARAM', 'error': 'algorithm is not supported'}
+
+        pepper = str(args['pepper'])
+        if pepper != self.lookup_pepper:
+            request.setResponseCode(400)
+            return {'errcode': 'M_INVALID_PEPPER', 'error': "pepper does not match '%s'" %
+                                                            self.lookup_pepper}
+
+        logger.info("Lookup of %d threepid(s) with algorithm %s", len(addresses), algorithm)
+        if algorithm == "none":
+            # Lookup without hashing
+            medium_address_tuples = []
+            for address_and_medium in addresses:
+                # Parse medium, address components
+                # The address and medium are flipped from what getMxids() is
+                # expecting, so switch them around
+                address, medium = address_and_medium.rsplit(maxsplit=1)
+
+                # Forbid addresses that contain a space
+                if " " in address:
+                    return {
+                        'errcode': 'M_UNKNOWN',
+                        'error': '"%s": contains spaces' % address
+                    }
```

Comment:
Need to set a response code.

Isn't there already a way to raise an exception to return error responses?

Reply:
I don't think so actually...

Reply:
Oh well...